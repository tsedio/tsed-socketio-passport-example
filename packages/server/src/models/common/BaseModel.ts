import {Immutable, ObjectID} from "@tsed/mongoose";
import {Example, Groups, Name, ReadOnly, Required} from "@tsed/schema";

export class HistoryBaseModel {
  @Required()
  @Immutable()
  @ReadOnly()
  @Name("created_at")
  @Groups("!create", "!update", "!summary", "!api")
  @Example(1667563214098)
  createdAt: number = Date.now();

  @Required()
  @ReadOnly()
  @Name("updated_at")
  @Groups("!create", "!update", "!summary", "!api")
  @Example(1667563214098)
  updatedAt: number = Date.now();
}

export class BaseModel extends HistoryBaseModel {
  @ObjectID("id")
  @Required()
  @Groups("!create")
  _id: string;
}
