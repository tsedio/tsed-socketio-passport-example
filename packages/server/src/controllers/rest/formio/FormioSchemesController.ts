import {Controller, Get, PathParams, QueryParams} from "@tsed/common";
import {Name, Returns, Summary} from "@tsed/schema";
import {getFormioSchema} from "@tsed/schema-formio";

@Controller("/formio")
@Name("Formio")
export class FormioSchemesController {
  @Get("/:alias")
  @Summary("Get schema for a model")
  @Returns(200, Object).ContentType("application/json").Header("Cache-Control", "no-cache, no-store, private, max-age=0")
  async getSchema(
    @PathParams("alias") alias: string,
    @QueryParams("groups", String) groups: string[]
  ) {
    const opts = {
      groups: [...(groups || []), "formio"],
      customKeys: true
    };

    return getFormioSchema(alias, opts);
  }
}
