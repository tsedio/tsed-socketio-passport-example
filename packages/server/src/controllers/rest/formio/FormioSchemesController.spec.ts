import { ApiKey, FormSchema } from "@clubmed/domain";
import { AuthTokenService, FormsClient, PermissionService } from "@clubmed/infra";
import { DataCompletionFormService } from "@clubmed/usecases";
import { PlatformTest } from "@tsed/common";
import { catchAsyncError, nameOf } from "@tsed/core";
import { NotFound } from "@tsed/exceptions";

import { FormioSchemesController } from "./FormioSchemesController";

async function getControllerFixture({ scopes = ["scope"], isAdmin = false } = {}) {
  const permissionService = {
    getScopes: jest.fn().mockReturnValue(scopes)
  };

  const formsClient = {
    getByName: jest.fn().mockResolvedValue(undefined),
    save: jest.fn().mockImplementation((o) => o),
    remove: jest.fn().mockResolvedValue(undefined)
  };

  const authTokenService = {
    claim: jest.fn().mockResolvedValue({
      permissions: ["clients:oidc"],
      isAdmin: jest.fn().mockReturnValue(isAdmin)
    })
  };

  const instance = await PlatformTest.invoke<FormioSchemesController>(FormioSchemesController, [
    {
      token: PermissionService,
      use: permissionService
    },
    {
      token: FormsClient,
      use: formsClient
    },
    {
      token: AuthTokenService,
      use: authTokenService
    }
  ]);

  return { instance, permissionService, formsClient };
}

describe("FormioFormioSchemesController", () => {
  beforeEach(() => PlatformTest.create({ cache: false }));
  afterEach(PlatformTest.reset);

  describe("getSchema", () => {
    it("should return OIDCClient Schema for creation", async () => {
      const { instance } = await getControllerFixture();

      // WHEN
      const result = await instance.getSchema("OIDCClient", ["create", "api_admin"]);

      // THEN
      expect(result).toMatchSnapshot();
    });
    it("should return ApiKey Schema with locales dataSources", async () => {
      const { instance } = await getControllerFixture({ scopes: ["api_admin"], isAdmin: true });

      // WHEN
      const result = await instance.getSchema(nameOf(ApiKey), []);

      // THEN
      expect(result).toMatchSnapshot();
    });
    it("should return the data completion form", async () => {
      const dataCompletionFormService = {
        getForm: jest.fn().mockReturnValue({ type: "form" })
      };
      const instance = await PlatformTest.invoke<FormioSchemesController>(FormioSchemesController, [
        {
          token: DataCompletionFormService,
          use: dataCompletionFormService
        },
        {
          token: AuthTokenService,
          use: {
            claim: jest.fn().mockResolvedValue({
              permissions: [],
              isAdmin: jest.fn().mockReturnValue(true)
            })
          }
        }
      ]);

      // WHEN
      const result = await instance.getSchema("DataCompletionRules", [], undefined, undefined, undefined, undefined, "routeId");

      // THEN
      expect(result).toEqual({ type: "form", permissions: ["api_admin"] });
      expect(dataCompletionFormService.getForm).toHaveBeenCalledWith("routeId", "production");
    });
    it("should return schema from database", async () => {
      const { instance, formsClient } = await getControllerFixture();
      const form = new FormSchema();
      form.name = "my_model";

      formsClient.getByName.mockReturnValue(form);

      // WHEN
      const result = await instance.getSchema("my_model", []);

      // THEN
      expect(result).toEqual(form);
    });
    it("should throw not found if model not found", async () => {
      const { instance } = await getControllerFixture();

      // WHEN
      let actualError;
      try {
        await instance.getSchema("ModelNotFound", ["create"]);
      } catch (err) {
        actualError = err;
      }

      // THEN
      expect(actualError).toBeInstanceOf(NotFound);
    });
    it("should merge form with another with on", async () => {
      const { instance } = await getControllerFixture();

      // WHEN
      const result = await instance.getSchema("quable_event", ["create", "api_admin"], undefined, undefined, "ApiConfig", "properties");

      // THEN
      expect(result).toMatchInlineSnapshot(`
        {
          "access": [],
          "components": [
            {
              "calculateValue": "value = _.snakeCase(value)",
              "description": "The unique configuration ID in the ClubMed API",
              "disabled": false,
              "input": true,
              "key": "id",
              "label": "Id:",
              "tableView": true,
              "type": "textfield",
              "validate": {
                "required": true,
              },
            },
            {
              "description": "Fill a description to explain the configuration usage in the ClubMed API",
              "disabled": false,
              "input": true,
              "key": "description",
              "label": "Description:",
              "tableView": false,
              "type": "textfield",
              "validate": {
                "required": true,
              },
            },
            {
              "components": [
                {
                  "disabled": false,
                  "input": true,
                  "key": "active",
                  "label": "Is active configuration:",
                  "type": "checkbox",
                  "validate": {
                    "required": false,
                  },
                },
                {
                  "data": {
                    "url": "/locales?include_wildcard=true",
                  },
                  "dataSrc": "url",
                  "disabled": false,
                  "idPath": "id",
                  "input": true,
                  "key": "locales_active",
                  "label": "Locales active:",
                  "multiple": true,
                  "selectThreshold": 0.3,
                  "template": "<span class="choices__item-template">{{item.id}}<small>{{item.businessUnit == "*" ? "include all locales" : item.businessUnit}}</small></span>",
                  "type": "select",
                  "validate": {
                    "min": 1,
                    "required": true,
                  },
                  "valueProperty": "id",
                  "widget": "choicesjs",
                },
                {
                  "components": [
                    {
                      "disabled": false,
                      "key": "day",
                      "label": "Day:",
                      "type": "radio",
                      "validate": {
                        "required": true,
                      },
                      "values": [
                        {
                          "label": "MONDAY",
                          "shortcut": "",
                          "value": "MONDAY",
                        },
                        {
                          "label": "TUESDAY",
                          "shortcut": "",
                          "value": "TUESDAY",
                        },
                        {
                          "label": "WEDNESDAY",
                          "shortcut": "",
                          "value": "WEDNESDAY",
                        },
                        {
                          "label": "THURSDAY",
                          "shortcut": "",
                          "value": "THURSDAY",
                        },
                        {
                          "label": "FRIDAY",
                          "shortcut": "",
                          "value": "FRIDAY",
                        },
                        {
                          "label": "SATURDAY",
                          "shortcut": "",
                          "value": "SATURDAY",
                        },
                        {
                          "label": "SUNDAY",
                          "shortcut": "",
                          "value": "SUNDAY",
                        },
                      ],
                    },
                    {
                      "components": [
                        {
                          "disabled": false,
                          "format": "HH:mm:ss",
                          "input": true,
                          "inputMask": "99:99:99",
                          "key": "opening_hour",
                          "label": "Opening hour:",
                          "type": "time",
                          "validate": {
                            "required": false,
                          },
                        },
                        {
                          "disabled": false,
                          "format": "HH:mm:ss",
                          "input": true,
                          "inputMask": "99:99:99",
                          "key": "closing_hour",
                          "label": "Closing hour:",
                          "type": "time",
                          "validate": {
                            "required": false,
                          },
                        },
                      ],
                      "disabled": false,
                      "input": true,
                      "key": "business_hours",
                      "label": "Business hours:",
                      "rowDrafts": false,
                      "type": "editgrid",
                      "validate": {
                        "required": true,
                      },
                    },
                  ],
                  "disabled": false,
                  "input": true,
                  "key": "weekdays",
                  "label": "Weekdays:",
                  "rowDrafts": false,
                  "type": "editgrid",
                  "validate": {
                    "min": 1,
                    "required": true,
                  },
                },
              ],
              "customClass": "p-5 pt-4 border-1 border-gray-200 rounded-small",
              "hideLabel": false,
              "input": true,
              "key": "properties",
              "label": "Properties",
              "tableView": false,
              "type": "container",
            },
          ],
          "display": "form",
          "machineName": "api-config",
          "name": "api-config",
          "permissions": [
            "clients:oidc",
            "authenticated",
          ],
          "submissionAccess": [],
          "tags": [],
          "title": "ApiConfig",
          "type": "form",
        }
      `);
    });
    it("should fallback to default merge_with given model if alias doesn't exists", async () => {
      const { instance } = await getControllerFixture();

      // WHEN
      const result = await instance.getSchema("unknown", ["create", "api_admin"], undefined, undefined, "ApiConfig", "properties");

      // THEN
      expect(result).toMatchInlineSnapshot(`
        {
          "access": [],
          "components": [
            {
              "calculateValue": "value = _.snakeCase(value)",
              "description": "The unique configuration ID in the ClubMed API",
              "disabled": false,
              "input": true,
              "key": "id",
              "label": "Id:",
              "tableView": true,
              "type": "textfield",
              "validate": {
                "required": true,
              },
            },
            {
              "description": "Fill a description to explain the configuration usage in the ClubMed API",
              "disabled": false,
              "input": true,
              "key": "description",
              "label": "Description:",
              "tableView": false,
              "type": "textfield",
              "validate": {
                "required": true,
              },
            },
            {
              "disabled": false,
              "input": true,
              "key": "properties",
              "label": "Properties",
              "tableView": false,
              "type": "datamap",
              "validate": {
                "required": false,
              },
              "valueComponent": {
                "hideLabel": true,
                "tableView": true,
              },
            },
          ],
          "display": "form",
          "machineName": "api-config",
          "name": "api-config",
          "permissions": [
            "clients:oidc",
            "authenticated",
          ],
          "submissionAccess": [],
          "tags": [],
          "title": "ApiConfig",
          "type": "form",
        }
      `);
    });
  });
  describe("createSchema()", () => {
    it("should create schema", async () => {
      const { instance } = await getControllerFixture();

      const form = new FormSchema();
      form.name = "test";

      // WHEN
      const result = await instance.createSchema(form);

      // THEN
      expect(result).toEqual(form);
    });
    it("should throw error if the name is already used by a schema", async () => {
      const { instance, formsClient } = await getControllerFixture();

      formsClient.getByName.mockResolvedValue(new FormSchema());

      const form = new FormSchema();
      form.name = "test";

      // WHEN
      const error = await catchAsyncError(() => instance.createSchema(form));

      // THEN
      expect(error.message).toEqual("Form name already exists");
    });
  });
  describe("updateSchema()", () => {
    it("should update schema", async () => {
      const { instance, formsClient } = await getControllerFixture();
      formsClient.getByName.mockResolvedValue(new FormSchema());

      const form = new FormSchema();
      form.name = "test";

      // WHEN
      const result = await instance.updateSchema("test", form);

      // THEN
      expect(result).toEqual(form);
    });
    it("should throw error if the name doesn't exists", async () => {
      const { instance } = await getControllerFixture();

      const form = new FormSchema();
      form.name = "test";

      // WHEN
      const error = await catchAsyncError(() => instance.updateSchema("test", form));

      // THEN
      expect(error.message).toEqual("Form name not found");
    });
  });
  describe("deleteSchema()", () => {
    it("should delete schema", async () => {
      const { instance, formsClient } = await getControllerFixture();
      const form = new FormSchema();
      form._id = "id";

      formsClient.getByName.mockResolvedValue(form);

      // WHEN
      await instance.deleteSchema("test");

      // THEN
      expect(formsClient.getByName).toHaveBeenCalledWith("test");
      expect(formsClient.remove).toHaveBeenCalledWith("id");
    });
    it("should throw error if the name doesn't exists", async () => {
      const { instance } = await getControllerFixture();

      const form = new FormSchema();
      form.name = "test";

      // WHEN
      const error = await catchAsyncError(() => instance.deleteSchema("test"));

      // THEN
      expect(error.message).toEqual("Form name not found");
    });
  });
});
