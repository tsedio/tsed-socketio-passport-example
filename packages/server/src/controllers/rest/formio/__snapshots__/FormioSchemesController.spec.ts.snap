// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormioFormioSchemesController getSchema should return ApiKey Schema with locales dataSources 1`] = `
{
  "access": [],
  "components": [
    {
      "description": "Api_key provided by ClubMed API team",
      "disabled": false,
      "input": true,
      "key": "id",
      "label": "Api Key:",
      "tableView": true,
      "type": "textfield",
      "validate": {
        "required": true,
      },
    },
    {
      "disabled": false,
      "input": true,
      "key": "contacts",
      "label": "Contacts:",
      "multiple": true,
      "tableView": true,
      "type": "email",
      "validate": {
        "required": true,
      },
    },
    {
      "description": "Commercial profile ID associated to this ApiKey",
      "disabled": false,
      "input": true,
      "key": "profile",
      "label": "Profile:",
      "tableView": true,
      "type": "textfield",
      "validate": {
        "required": false,
      },
    },
    {
      "data": {
        "url": "/locales?include_wildcard=true",
      },
      "dataSrc": "url",
      "disabled": false,
      "idPath": "id",
      "input": true,
      "key": "locales",
      "label": "Locales:",
      "multiple": true,
      "selectThreshold": 0.3,
      "template": "<span class="choices__item-template">{{item.id}}<small>{{item.businessUnit == "*" ? "include all locales" : item.businessUnit}}</small></span>",
      "type": "select",
      "validate": {
        "required": true,
      },
      "valueProperty": "id",
      "widget": "choicesjs",
    },
    {
      "defaultValue": "TBD",
      "description": "Application code",
      "disabled": false,
      "input": true,
      "key": "application_code",
      "label": "Application code:",
      "type": "textfield",
      "validate": {
        "required": false,
      },
    },
    {
      "defaultValue": "TBD",
      "description": "The channel code used to call CLIC backend",
      "disabled": false,
      "input": true,
      "key": "clic_channel_code",
      "label": "Clic channel code:",
      "tableView": true,
      "type": "textfield",
      "validate": {
        "required": false,
      },
    },
    {
      "defaultValue": "NET",
      "description": "The channel code applied for a commercial transaction",
      "disabled": false,
      "input": true,
      "key": "channel_code",
      "label": "Channel code:",
      "tableView": true,
      "type": "textfield",
      "validate": {
        "required": false,
      },
    },
    {
      "defaultValue": true,
      "description": "Is the date offset should be considered for the api_key",
      "disabled": false,
      "input": true,
      "key": "date_offset_enabled",
      "label": "Enable date offset",
      "type": "checkbox",
      "validate": {
        "required": false,
      },
    },
    {
      "defaultValue": true,
      "description": "Is used",
      "disabled": true,
      "input": true,
      "key": "used",
      "label": "Used",
      "tableView": true,
      "type": "checkbox",
      "validate": {
        "required": false,
      },
    },
    {
      "defaultValue": true,
      "description": "Allow or disallow the access to API for this ApiKey",
      "disabled": false,
      "input": true,
      "key": "enabled",
      "label": "Enabled",
      "tableView": true,
      "type": "checkbox",
      "validate": {
        "required": false,
      },
    },
    {
      "description": "The clubmed agent id",
      "disabled": false,
      "input": true,
      "key": "agent_id",
      "label": "Agent id",
      "type": "textfield",
      "validate": {
        "required": false,
      },
    },
  ],
  "display": "form",
  "machineName": "api-key",
  "name": "api-key",
  "permissions": [
    "clients:oidc",
    "api_admin",
  ],
  "submissionAccess": [],
  "tags": [],
  "title": "ApiKey",
  "type": "form",
}
`;

exports[`FormioFormioSchemesController getSchema should return OIDCClient Schema for creation 1`] = `
{
  "access": [],
  "components": [
    {
      "disabled": true,
      "hidden": true,
      "input": true,
      "key": "created",
      "label": "Created:",
      "type": "textfield",
      "validate": {
        "required": false,
      },
    },
    {
      "disabled": true,
      "hidden": true,
      "input": true,
      "key": "modified",
      "label": "Modified:",
      "type": "textfield",
      "validate": {
        "required": false,
      },
    },
    {
      "components": [
        {
          "components": [
            {
              "disabled": false,
              "input": true,
              "key": "client_name",
              "label": "Client name:",
              "type": "textfield",
              "validate": {
                "required": true,
              },
            },
            {
              "defaultValue": false,
              "disabled": false,
              "input": true,
              "key": "trusted",
              "label": "Trusted",
              "type": "checkbox",
              "validate": {
                "required": false,
              },
            },
            {
              "disabled": false,
              "input": true,
              "key": "client_uri",
              "label": "Client Uri:",
              "type": "url",
              "validate": {
                "required": false,
              },
            },
            {
              "disabled": false,
              "input": true,
              "key": "logo_uri",
              "label": "Logo uri:",
              "type": "url",
              "validate": {
                "required": false,
              },
            },
            {
              "defaultValue": "web",
              "disabled": false,
              "key": "application_type",
              "label": "Application type:",
              "type": "radio",
              "validate": {
                "required": true,
              },
              "values": [
                {
                  "label": "web",
                  "shortcut": "",
                  "value": "web",
                },
                {
                  "label": "native",
                  "shortcut": "",
                  "value": "native",
                },
                {
                  "label": "service",
                  "shortcut": "",
                  "value": "service",
                },
              ],
            },
            {
              "disabled": false,
              "inline": false,
              "key": "response_types",
              "label": "Response types:",
              "type": "selectboxes",
              "validate": {
                "required": false,
              },
              "values": [
                {
                  "label": "code",
                  "shortcut": "",
                  "value": "code",
                },
                {
                  "label": "id_token token",
                  "shortcut": "",
                  "value": "id_token token",
                },
                {
                  "label": "code id_token token",
                  "shortcut": "",
                  "value": "code id_token token",
                },
                {
                  "label": "token",
                  "shortcut": "",
                  "value": "token",
                },
                {
                  "label": "none",
                  "shortcut": "",
                  "value": "none",
                },
              ],
            },
            {
              "disabled": false,
              "inline": false,
              "key": "grant_types",
              "label": "Grant types:",
              "type": "selectboxes",
              "validate": {
                "required": false,
              },
              "values": [
                {
                  "label": "authorization_code",
                  "shortcut": "",
                  "value": "authorization_code",
                },
                {
                  "label": "implicit",
                  "shortcut": "",
                  "value": "implicit",
                },
                {
                  "label": "refresh_token",
                  "shortcut": "",
                  "value": "refresh_token",
                },
                {
                  "label": "client_credentials",
                  "shortcut": "",
                  "value": "client_credentials",
                },
              ],
            },
            {
              "defaultValue": "client_secret_basic",
              "disabled": false,
              "key": "token_endpoint_auth_method",
              "label": "Token endpoint authentication method:",
              "type": "radio",
              "validate": {
                "required": false,
              },
              "values": [
                {
                  "label": "client_secret_post",
                  "shortcut": "",
                  "value": "client_secret_post",
                },
                {
                  "label": "client_secret_basic",
                  "shortcut": "",
                  "value": "client_secret_basic",
                },
                {
                  "label": "client_secret_jwt",
                  "shortcut": "",
                  "value": "client_secret_jwt",
                },
                {
                  "label": "private_key_jwt",
                  "shortcut": "",
                  "value": "private_key_jwt",
                },
                {
                  "label": "none",
                  "shortcut": "",
                  "value": "none",
                },
              ],
            },
            {
              "defaultValue": 3600,
              "delimiter": false,
              "disabled": false,
              "input": true,
              "inputFormat": "plain",
              "key": "default_max_age",
              "label": "Default max age:",
              "mask": false,
              "requireDecimal": false,
              "type": "number",
              "validate": {
                "required": false,
              },
            },
          ],
          "key": "parameters",
          "label": "Parameters",
        },
        {
          "components": [
            {
              "disabled": false,
              "input": true,
              "key": "redirect_uris",
              "label": "Authorized redirect Uris:",
              "multiple": true,
              "type": "textfield",
              "validate": {
                "required": false,
              },
            },
            {
              "disabled": false,
              "input": true,
              "key": "post_logout_redirect_uris",
              "label": "Authorized post logout redirect Uris:",
              "multiple": true,
              "type": "textfield",
              "validate": {
                "required": false,
              },
            },
          ],
          "key": "redirectUris",
          "label": "Redirect Uris",
        },
        {
          "components": [
            {
              "customConditional": "show = form?.permissions?.includes("api_admin")",
              "data": {
                "url": "/oidc/scopes?",
              },
              "dataSrc": "url",
              "disabled": false,
              "idPath": "name",
              "input": true,
              "key": "scopes",
              "label": "Scopes:",
              "multiple": true,
              "selectThreshold": 0.3,
              "template": "<span>{{ item.name }}</span>",
              "type": "select",
              "validate": {
                "required": false,
              },
              "valueProperty": "name",
              "widget": "choicesjs",
            },
          ],
          "key": "permissions",
          "label": "Permissions",
        },
      ],
      "input": false,
      "key": "tabs",
      "label": "Tabs",
      "pushed": true,
      "tableView": false,
      "type": "tabs",
    },
    {
      "disabled": false,
      "hidden": true,
      "input": true,
      "key": "type",
      "label": "Type",
      "type": "textfield",
      "validate": {
        "required": false,
      },
    },
  ],
  "display": "form",
  "machineName": "oidc-client",
  "name": "oidc-client",
  "permissions": [
    "clients:oidc",
    "authenticated",
  ],
  "submissionAccess": [],
  "tags": [],
  "title": "OIDCClient",
  "type": "form",
}
`;
